src/cairoUtilFuncGen/abi/abiDecode.ts

```diff
@@ -62,47 +64,65 @@ export class AbiDecode extends StringIndexedFuncGenWithAuxiliar {
     );
     const typesToDecode = types instanceof TupleType ? types.elements : [types];
 
-    const functionName = this.getOrCreate(typesToDecode.map((t) => generalizeType(t)[0]));
+    const generatedFunction = this.getOrCreateFuncDef(
+      typesToDecode.map((t) => generalizeType(t)[0]),
+    );
+
+    return createCallToFunction(generatedFunction, [expressions[0]], this.ast);
+  }
+
+  public getOrCreateFuncDef(types: TypeNode[]): CairoFunctionDefinition {
+    const key = types.map((t) => t.pp()).join(',');
+    const existing = this.generatedFunctionsDef.get(key);
+    if (existing !== undefined) {
+      return existing;
+    }
+
+    const funcInfo = this.getOrCreate(types);
 
-    const functionStub = createCairoFunctionStub(
-      functionName,
+    const funcDef = createCairoGeneratedFunction(
+      funcInfo,
       [['data', createBytesTypeName(this.ast), DataLocation.Memory]],
-      typesToDecode.map((t, index) =>
+      types.map((t, index) =>
         isValueType(t)
           ? [`result${index}`, typeNameFromTypeNode(t, this.ast)]
           : [`result${index}`, typeNameFromTypeNode(t, this.ast), DataLocation.Memory],
       ),
-      ['bitwise_ptr', 'range_check_ptr', 'warp_memory'],
       this.ast,
-      sourceUnit ?? this.sourceUnit,
+      this.sourceUnit,
     );
 
-    return createCallToFunction(functionStub, [expressions[0]], this.ast);
+    this.generatedFunctionsDef.set(key, funcDef);
+    return funcDef;
```

Comment:
Usually when you've got a `getOrCreate` its good to have a `get` and `create` methods separately too

Reply:
Yes, I agree. Can we postpone that? It will mean changing all the function generators :cold_sweat:

Reply:
Sure